### PR TITLE
feat: verify token with public key

### PR DESCRIPTION
**Pull request by koonpeng**
Stylistic changes are a result of running uncrustify with rmf's style.

Previously it is able to use `pubkey` option to let soss-websocket use a public key, but it is "broken" in that it expects the key in raw string form, this changed it so that the `pubkey` option refers to a pem-encoded public key file.

This also adds some docs on the configuration options for soss-websocket and improve on error messages when token verification fails.

### CODE REVIEW

File: packages/websocket/src/JwtValidator.cpp [hunk 1]
```diff
@@ -1,74 +1,98 @@
...
 namespace websocket {
 
-bool JwtValidator::verify(const std::string& token)
+VerificationPolicy::VerificationPolicy(std::vector<Rule> rules,
+  std::vector<Rule> header_rules, std::string secret_or_pubkey)
+: _secret_or_pubkey(std::move(secret_or_pubkey)), _rules(std::move(rules)),
+  _header_rules(std::move(header_rules))
+{
+  // This is so that we don't have to create the regexes everytime the policy is used.
+  for (auto& r : _rules)
```
Comment (mxgrey) [inline packages/websocket/src/JwtValidator.cpp]:
```suggestion
  for (const auto& r : _rules)
```

File: packages/websocket/src/JwtValidator.cpp [hunk 2]
```diff
@@ -1,74 +1,98 @@
...
+    _matchers[r.first] = std::regex{r.second};
+  for (auto& r : _header_rules)
```
Comment (mxgrey) [inline packages/websocket/src/JwtValidator.cpp]:
```suggestion
  for (const auto& r : _header_rules)
```

File: packages/websocket/src/JwtValidator.cpp [hunk 3]
```diff
@@ -1,74 +1,98 @@
...
 #include <unordered_map>
 #include <regex>
+#include <utility>
+#include <iostream>
 
 namespace soss {
 namespace websocket {
 
-bool JwtValidator::verify(const std::string& token)
+VerificationPolicy::VerificationPolicy(std::vector<Rule> rules,
```
Comment (mxgrey) [inline packages/websocket/src/JwtValidator.cpp]: The argument formatting is kind of a mess here.

File: packages/websocket/src/JwtValidator.cpp [hunk 4]
```diff
@@ -1,74 +1,98 @@
...
+
+void VerificationPolicy::check(const std::string& token,
+  const json_t& header,
+  const json_t& payload)
+{
+  jwt::decode(token, jwt::params::algorithms(
+      {header["alg"].get_ref<const std::string&>()}), jwt::params::secret(
+      _secret_or_pubkey));
+
+  for (auto& r : _header_rules)
```
Comment (mxgrey) [inline packages/websocket/src/JwtValidator.cpp]:
```suggestion
  for (const auto& r : _header_rules)
```

File: packages/websocket/src/JwtValidator.cpp [hunk 5]
```diff
@@ -1,74 +1,98 @@
...
+    if (it == header.end())
+      throw jwt::VerificationError("'" + r.first + "' not found in headers");
+    if (!it->is_string())
+      throw jwt::VerificationError("'" + r.first + "' expected to be string");
+    const auto& s = it->get_ref<const std::string&>();
+    if (!std::regex_match(s, _header_matchers.at(r.first)))
+      throw jwt::VerificationError("'" + r.first + "' does not match policy");
+  }
+
+  for (auto& r : _rules)
```
Comment (mxgrey) [inline packages/websocket/src/JwtValidator.cpp]:
```suggestion
  for (const auto& r : _rules)
```

File: packages/websocket/src/JwtValidator.cpp [hunk 6]
```diff
@@ -1,74 +1,98 @@
...
+  _header_rules(std::move(header_rules))
+{
+  // This is so that we don't have to create the regexes everytime the policy is used.
+  for (auto& r : _rules)
+    _matchers[r.first] = std::regex{r.second};
+  for (auto& r : _header_rules)
+    _header_matchers[r.first] = std::regex(r.second);
+}
+
+void VerificationPolicy::check(const std::string& token,
```
Comment (mxgrey) [inline packages/websocket/src/JwtValidator.cpp:29]:
```suggestion
void VerificationPolicy::check(
  const std::string& token,
```

File: packages/websocket/src/JwtValidator.hpp [hunk 7]
```diff
@@ -18,26 +18,50 @@
...
+class VerificationPolicy
 {
-  std::string secret_or_pub;
-  std::string algo;
-};
+public:
+
+  using Rule = std::pair<std::string, std::string>;
+
+  VerificationPolicy(std::vector<Rule> rules,
```
Comment (mxgrey) [inline packages/websocket/src/JwtValidator.hpp:34]:
```suggestion
  VerificationPolicy(
    std::vector<Rule> rules,
```

File: packages/websocket/src/JwtValidator.hpp [hunk 8]
```diff
@@ -18,26 +18,50 @@
...
+
+  inline const std::string& secret_or_pubkey() const
+  {
+    return _secret_or_pubkey;
+  }
+
+  /**
+   * @throws VerificationError
+   */
+  void check(const std::string& token, const json_t& header,
```
Comment (mxgrey) [inline packages/websocket/src/JwtValidator.hpp]:
```suggestion
  void check(
    const std::string& token,
    const json_t& header,
```

File: packages/websocket/src/ServerConfig.cpp [hunk 9]
```diff
@@ -29,74 +33,121 @@ const std::string YamlSecretKey = "secret";
...
-      secret_or_pub = policy_node[YamlSecretKey].as<std::string>();
     else
-      secret_or_pub = policy_node[YamlPubkeyKey].as<std::string>();
-
-    std::string algo = policy_node[YamlAlgoKey].as<std::string>();
-
-    jwt_validator.add_verification_policy(VerificationPolicies::match_all(
-      rules, secret_or_pub, algo));
+    {
+      for (auto& policy_node : policies_node)
```
Comment (mxgrey) [inline packages/websocket/src/ServerConfig.cpp]:
```suggestion
      for (const auto& policy_node : policies_node)
```

File: packages/websocket/src/ServerConfig.cpp [hunk 10]
```diff
@@ -29,74 +33,121 @@ const std::string YamlSecretKey = "secret";
...
-    {}, secret_or_pub, algo));
   return true;
 }
 
-std::string ServerConfig::glob_to_regex(const std::string& s)
+std::string ServerConfig::_glob_to_regex(const std::string& s)
 {
   using namespace boost::algorithm;
 
   return find_format_all_copy(s, token_finder(is_any_of(".*?\\")), [](auto s)
```
Comment (mxgrey) [inline packages/websocket/src/ServerConfig.cpp]:
```suggestion
  return find_format_all_copy(
    s, token_finder(is_any_of(".*?\\")), [](const auto& s)
```

File: packages/websocket/src/ServerConfig.cpp [hunk 11]
```diff
@@ -29,74 +33,121 @@ const std::string YamlSecretKey = "secret";
...
+  {
+    header_rules.emplace_back(VerificationPolicy::Rule("alg",
+      _glob_to_regex(policy_node[YamlAlgoKey].as<std::string>())));
+  }
+
+  for (const auto& r : policy_node[YamlRulesKey])
+  {
+    std::string regex_pattern = _glob_to_regex(r.second.as<std::string>());
+    rules.emplace_back(VerificationPolicy::Rule{
+        r.first.as<std::string>(), regex_pattern
```
Comment (mxgrey) [inline packages/websocket/src/ServerConfig.cpp]:
```suggestion
        r.first.as<std::string>(), std::move(regex_pattern)
```

File: packages/websocket/src/ServerConfig.cpp [hunk 12]
```diff
@@ -29,74 +33,121 @@ const std::string YamlSecretKey = "secret";
...
+      });
+  }
+
+  return VerificationPolicy(
+    rules, header_rules, secret_or_pubkey);
```
Comment (mxgrey) [inline packages/websocket/src/ServerConfig.cpp]:
```suggestion
    std::move(rules), std::move(header_rules), std::move(secret_or_pubkey));
```

**Review by mxgrey — CHANGES_REQUESTED** [review]
This looks great! I just left some nitpicks about formatting, const usage, and some move semantics.

**Comment by koonpeng** [comment]
Thanks! Fixed the formatting.

**Review by mxgrey — APPROVED** [review]

**Pull request merged**